### ISSUE TITLE
Inconsistent behavior for recursive types

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#2659]
Running `pyright 1.1.192`. 

There is inconsistent behavior for type-checking against a recursive type. A particular literal will be marked as valid, whereas a variable referencing the same literal will be marked as invalid, in the context of that recursive type. Here is an example:

```python
from typing import Union, List

S = Union[int, bool, List["S"]]

def g(x: S):
    pass

g(1)   # okay
g([1]) # okay

an_int = 1
a_list = [1]

g(an_int)   # okay
g(a_list)   # invalid
```
```
Argument of type "list[int]" cannot be assigned to parameter "x" of type "S" in function "g"
  Type "list[int]" cannot be assigned to type "S"
    "list[int]" is incompatible with "int"
    "list[int]" is incompatible with "bool"
    TypeVar "_T@list" is invariant
      Type "int" cannot be assigned to type "S"
        "int" is incompatible with "bool"
        "int" is incompatible with "List[S]"
```

**Comment by erictraut** [comment]
The behavior you're seeing is correct, although I can understand why it's somewhat counterintuitive.

The type parameter for `list` is invariant, so the type `list[S]` is not compatible with the type `list[int]`. When a list expression is used, it needs to be properly typed as `list[S]` from the beginning. In the example above, the list expression in the statement `g([1])` is able to "start life" as type `list[S]` because pyright uses bidirectional type inference to infer that it should be `list[S]` rather than `list[int]` or `list[Literal[1]]`. However, the statement `a_list = [1]` has no context, so pyright falls back to its normal type inference rules for list and infers the type to be `list[int]`.

Here are a couple of workarounds. The most straightforward workaround is to add a type annotation for the variable `a_list` so it "begins life" as a `list[S]`.

```python
a_list: list[S] = [1]
```

Another option is to change the definition of `S` to replace `list` with `Sequence`, which has a covariant type parameter.

**Comment by rsokl** [comment]
Ah, I see. I had never really thought about mutable generic containers needing to be invariant. This throws quite a wrench in my plans -- I wanted to provide my users with very-precise annotations of what lists/dicst/sets are permitted for a serialization process -- but I certainly see why this is as-designed. Thanks for the clear explanation!
